src/uct/ugni/smsg/ugni_smsg_ep.c

```diff
@@ -0,0 +1,268 @@
+UCS_CLASS_DEFINE_DELETE_FUNC(uct_ugni_smsg_ep_t, uct_ep_t);
+
+ucs_status_t uct_ugni_smsg_ep_get_address(uct_ep_h tl_ep, struct sockaddr *addr) {
+
+  uct_ugni_smsg_ep_t *ep = ucs_derived_of(tl_ep, uct_ugni_smsg_ep_t);
+  uct_sockaddr_smsg_ugni_t *ep_addr = (uct_sockaddr_smsg_ugni_t*)addr;
+
+  ucs_status_t rc;
+
+  rc = uct_ugni_iface_get_address(tl_ep->iface, addr);
+
+  if(rc != UCS_OK){
+    return rc;
+  }
+
+  ep_addr->ep_hash = ep->super.hash_key;
+  memcpy(&ep_addr->smsg_attr, &ep->smsg_attr->mbox_attr, sizeof(gni_smsg_attr_t));
+
+  return UCS_OK;
+}
+
+ucs_status_t uct_ugni_smsg_ep_connect_to_ep(uct_ep_h tl_ep, const struct sockaddr *addr){
+  uct_ugni_smsg_ep_t *ep = ucs_derived_of(tl_ep, uct_ugni_smsg_ep_t);
+  uct_ugni_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_ugni_iface_t);
+  const uct_sockaddr_smsg_ugni_t *iface_addr = (const uct_sockaddr_smsg_ugni_t*)addr;
+  gni_smsg_attr_t *local_attr = (gni_smsg_attr_t*)&ep->smsg_attr->mbox_attr;
+  gni_smsg_attr_t *remote_attr = (gni_smsg_attr_t *)&iface_addr->smsg_attr;
+  gni_return_t gni_rc;
+  ucs_status_t rc = UCS_OK;
+
+  pthread_mutex_lock(&uct_ugni_global_lock);
+  rc = ugni_connect_ep(iface, &iface_addr->super, &ep->super);
+
+  if(rc != UCS_OK){
+    ucs_error("Could not connect ep in smsg");
+    goto exit_lock;
+  }
+
+  gni_rc = GNI_SmsgInit(ep->super.ep, local_attr, remote_attr);
+
+  if(GNI_RC_SUCCESS != gni_rc){
+    ucs_error("Failed to initalize smsg. %s [%i]", gni_err_str[gni_rc], gni_rc);
+    if(GNI_RC_INVALID_PARAM == gni_rc){
+      rc = UCS_ERR_INVALID_PARAM;
+    } else {
+      rc = UCS_ERR_NO_MEMORY;
+    }
+    goto exit_lock;
+  }
+
+  uint32_t the_hash = (uint32_t)iface_addr->ep_hash;
+  gni_rc = GNI_EpSetEventData(ep->super.ep, iface->domain_id, the_hash);
+
+  if(GNI_RC_SUCCESS != gni_rc){
+    ucs_error("Could not set GNI_EpSetEventData!");
+  }
+
+
+ exit_lock:
+  pthread_mutex_unlock(&uct_ugni_global_lock);
+
+  return rc;
+}
+
+/* A common mm active message sending function.
+ * The first parameter indicates the origin of the call.
+ * 1 - perform am short sending
+ * 0 - perform am bcopy sending
+ */
+
+static UCS_F_ALWAYS_INLINE ssize_t
+uct_ugni_smsg_ep_am_common_send(uct_ugni_smsg_ep_t *ep, uct_ugni_smsg_iface_t *iface,
+                                uint8_t am_id, unsigned header_length, void *header,
+                                unsigned payload_length, void *payload, uct_ugni_smsg_desc_t *desc)
+{
+  gni_return_t gni_rc;
+
+  desc->msg_id = iface->smsg_id++;
+  desc->ep = &ep->super;
+
+  gni_rc = GNI_SmsgSendWTag(ep->super.ep, header, header_length, payload, payload_length, desc->msg_id, am_id);
+
+  if(gni_rc != GNI_RC_SUCCESS){
+    /* look for better return code. Maybe steak pasha's macro */
+    ucs_mpool_put(desc);
+    return UCS_ERR_NO_RESOURCE;
+  }
+
+  ++ep->super.outstanding;
+  ++iface->super.outstanding;
+
+  sglib_hashed_uct_ugni_smsg_desc_t_add(iface->smsg_list, desc);
```

Comment:
Honestly, that may have been left over when I had a worse return code and didn't fix the comment when I fixed that.